### PR TITLE
Update finagle and other cosmos dependencies

### PR DESCRIPTION
**Pull request by takirala**
This PR upgrades following dependencies in cosmos : 

```
-  "com.twitter.common" % "util-system-mocks" % "0.0.27" -> % "0.0.105",
-  "com.twitter.common" % "quantity" % "0.0.31" -> "0.0.100"
-  val bijection = "0.9.4" -> "0.9.5"
-  val circe = "0.6.1" -> "0.8.0"
-  val finch = "0.11.1" -> "0.15.1"
-  val twitterServer = "1.25.0" -> "1.30.0"
-  val twitterUtilCore = "6.39.0" -> "6.45.0"
```

Dependency Upgrades in second go : 
```
- "com.google.code.findbugs" % "jsr305" % "3.0.1" -> "3.0.2" 
- Added "io.netty" % "netty-common" % V.netty (V.netty is 4.1.13.Final)
- val aws = "1.11.63" -> "1.11.163"
- val curator = "2.11.1" -> "2.12.0"
- val fastparse = "0.4.1" -> "0.4.3"
- val guava = "16.0.1" -> "22.0"
- val logback = "1.1.3" -> "1.2.3"
- val mockito = "1.10.19" -> "2.8.47"
- val mustache = "0.9.1" -> "0.9.5"
- val scalaUri = "0.4.11" -> "0.4.16"
- val slf4j = "1.7.10" -> "1.7.25"
```

Dependency List of cosmos before this upgrade : 

Original Dependency List

[Latest Ivy Dependency Report](https://github.com/dcos/cosmos/files/1166553/report.zip)

### CODE REVIEW

**Comment by takirala** [comment]
We require the `6.39.0` version of twitter util core as it is used by `bijection util` which is at `0.9.5` (its latest version).

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/Services.scala [hunk 1]
```diff
@@ -59,16 +55,9 @@ object Services {
         case false =>
           Http.client
         case true =>
-          Http.client
-            .configured(Transport.TLSClientEngine(Some({
-              case inet: InetSocketAddress => Ssl.client(hostname, inet.getPort)
-              case _ => Ssl.client()
-            })))
-            .configured(Transporter.TLSHostname(Some(hostname)))
+          Http.client.withTls(hostname)
```
Comment (takirala) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/Services.scala]: I tried this : 

```
Http.client
            .configured(new Param(new ConstClientEngineFactory(_ =>
              SslClientEngineFactory.createEngine(
                SSLContext.getDefault(),
                Address(new InetSocketAddress(hostname, port)),
                SslClientConfiguration(Some(hostname))
              )
            )))
```
and got this : 

```
 - be able to connect to an https site *** FAILED ***
[info]     com.mesosphere.cosmos.error.CosmosException: Unable to complete request due to service [adminRouter] unavailability
[info]     at com.mesosphere.cosmos.error.CosmosException$.apply(CosmosException.scala:27)
[info]     at com.mesosphere.cosmos.Services$ConnectionExceptionHandler$$anonfun$apply$1.applyOrElse(Services.scala:95)
[info]     at com.mesosphere.cosmos.Services$ConnectionExceptionHandler$$anonfun$apply$1.applyOrElse(Services.scala:93)
[info]     at com.twitter.util.Future$$anonfun$rescue$1.apply(Future.scala:1159)
[info]     at com.twitter.util.Future$$anonfun$rescue$1.apply(Future.scala:1157)
[info]     at com.twitter.util.Promise$Transformer.liftedTree1$1(Promise.scala:226)
[info]     at com.twitter.util.Promise$Transformer.k(Promise.scala:226)
[info]     at com.twitter.util.Promise$Transformer.apply(Promise.scala:236)
[info]     at com.twitter.util.Promise$Transformer.apply(Promise.scala:217)
[info]     at com.twitter.util.Promise$$anon$7.run(Promise.scala:530)
[info]     ...
[info]     Cause: com.twitter.finagle.ChannelClosedException: Connection reset by peer at remote address: www.google.com/172.217.5.100:443 from service: adminRouter. Remote Info: Upstream Address: Not Available, Upstream Client Id: Not Available, Downstream Address: www.google.com/172.217.5.100:443, Downstream Client Id: adminRouter, Trace Id: b031eeaf27654919.b031eeaf27654919<:b031eeaf27654919
[info]     at com.twitter.finagle.ChannelException$.apply(Exceptions.scala:255)
[info]     at com.twitter.finagle.netty4.transport.ChannelTransport$$anon$1.exceptionCaught(ChannelTransport.scala:190)
[info]     at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
[info]     at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
[info]     at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
[info]     at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
[info]     at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
[info]     at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
[info]     at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
[info]     at com.twitter.finagle.netty4.http.handler.ClientExceptionMapper$.exceptionCaught(ClientExceptionMapper.scala:38)
[info]     ...
[info]     Cause: java.io.IOException: Connection reset by peer
[info]     at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
[info]     at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
[info]     at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
[info]     at sun.nio.ch.IOUtil.read(IOUtil.java:192)
[info]     at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
[info]     at io.netty.buffer.UnpooledUnsafeDirectByteBuf.setBytes(UnpooledUnsafeDirectByteBuf.java:433)
[info]     at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1100)
[info]     at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:372)
[info]     at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123)
[info]     at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
[info]     ...
[info] ScalaTest
```

Release notes - https://github.com/twitter/finagle/releases/tag/finagle-6.44.0

File: cosmos-common/src/main/scala/com/mesosphere/cosmos/finch/FinchExtensions.scala [hunk 2]
```diff
@@ -31,7 +31,7 @@ object FinchExtensions {
   )(
     requestValidator: Endpoint[EndpointContext[Req, Res]]
   ): Endpoint[Json] = {
-    (base :: requestValidator).apply((context: EndpointContext[Req, Res]) => handler(context))
+    (base :: requestValidator).mapOutputAsync[Json] { case context => handler(context) }
```
Comment (cruhland) [inline cosmos-common/src/main/scala/com/mesosphere/cosmos/finch/FinchExtensions.scala]: Any expression of the form `x => f(x)` can always be simplified to `f` (via [eta-conversion](https://en.wikipedia.org/wiki/Lambda_calculus#.CE.B7-conversion)), so this line can be written:

```scala
(base :: requestValidator).mapOutputAsync[Json](handler)
```

Not sure if the `[Json]` part is necessary either.

**Comment by takirala** [comment]
@jsancio @cruhland @larioj 
During bootstrap process, this error is thrown : 

```
[Step 4/5] [info] <<cosmos-server>> java.lang.ClassNotFoundException: jdk.internal.misc.Unsafe
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.simontuffs.onejar.JarClassLoader.findClass(JarClassLoader.java:713) ~[cosmos-server_2.11-0.5.0-SNAPSHOT-one-jar.jar:na]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_51]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.simontuffs.onejar.JarClassLoader.loadClass(JarClassLoader.java:630) ~[cosmos-server_2.11-0.5.0-SNAPSHOT-one-jar.jar:na]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_51]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.internal.PlatformDependent0$6.run(PlatformDependent0.java:295) ~[netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at java.security.AccessController.doPrivileged(Native Method) ~[na:1.8.0_51]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.internal.PlatformDependent0.<clinit>(PlatformDependent0.java:288) ~[netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.internal.PlatformDependent.getSystemClassLoader(PlatformDependent.java:887) [netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.internal.PlatformDependent.isAndroid0(PlatformDependent.java:911) [netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.internal.PlatformDependent.<clinit>(PlatformDependent.java:70) [netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.ResourceLeakDetector.<init>(ResourceLeakDetector.java:155) [netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.ResourceLeakDetector.<init>(ResourceLeakDetector.java:200) [netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.ResourceLeakDetectorFactory$DefaultResourceLeakDetectorFactory.newResourceLeakDetector(ResourceLeakDetectorFactory.java:201) [netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at io.netty.util.HashedWheelTimer.<clinit>(HashedWheelTimer.java:89) [netty-common-4.1.10.Final.jar:4.1.10.Final]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.finagle.netty4.Netty4HashedWheelTimer.<init>(Netty4HashedWheelTimer.scala:100) [finagle-netty4_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.finagle.netty4.Netty4HashedWheelTimer$.<init>(Netty4HashedWheelTimer.scala:113) [finagle-netty4_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.finagle.netty4.Netty4HashedWheelTimer$.<clinit>(Netty4HashedWheelTimer.scala) [finagle-netty4_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.finagle.Http$Client$.<init>(Http.scala:178) [finagle-http_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.finagle.Http$Client$.<clinit>(Http.scala) [finagle-http_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.finagle.Http$.<init>(Http.scala:358) [finagle-http_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.finagle.Http$.<clinit>(Http.scala) [finagle-http_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.server.AdminHttpServer$class.com$twitter$server$AdminHttpServer$$startServer(AdminHttpServer.scala:216) [twitter-server_2.11-1.30.0.jar:1.30.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.server.AdminHttpServer$$anonfun$1.apply$mcV$sp(AdminHttpServer.scala:233) [twitter-server_2.11-1.30.0.jar:1.30.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.app.App$$anonfun$nonExitingMain$2.apply(App.scala:216) [util-app_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.app.App$$anonfun$nonExitingMain$2.apply(App.scala:216) [util-app_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59) [scala-library-2.11.7.jar:na]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48) [scala-library-2.11.7.jar:na]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.app.App$class.nonExitingMain(App.scala:216) [util-app_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.mesosphere.cosmos.Cosmos$.nonExitingMain(Cosmos.scala:390) [cosmos-server_2.11-0.5.0-SNAPSHOT.jar:0.5.0-SNAPSHOT]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.twitter.app.App$class.main(App.scala:190) [util-app_2.11-6.45.0.jar:6.45.0]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.mesosphere.cosmos.Cosmos$.main(Cosmos.scala:390) [cosmos-server_2.11-0.5.0-SNAPSHOT.jar:0.5.0-SNAPSHOT]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.mesosphere.cosmos.Cosmos.main(Cosmos.scala) [cosmos-server_2.11-0.5.0-SNAPSHOT.jar:0.5.0-SNAPSHOT]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_51]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_51]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_51]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_51]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.simontuffs.onejar.Boot.run(Boot.java:340) [cosmos-server_2.11-0.5.0-SNAPSHOT-one-jar.jar:na]
[12:44:23][Step 4/5] [info] <<cosmos-server>> 	at com.simontuffs.onejar.Boot.main(Boot.java:166) [cosmos-server_2.11-0.5.0-SNAPSHOT-one-jar.jar:na]
```

This issue comes from netty and is discussed [here](https://github.com/netty/netty/issues/6855). The github issue comments suggest that netty should be upgraded to `4.1.12.Final` and this won't happen. However we have already hit the latest version of modules which use netty in their dependencies and they don't (yet) use the `4.1.12.Final` As I don't see any performance affect, can we live with this exception during bootstrap or should I investigate further ?

**Comment by cruhland** [comment]
@takirala what version of Netty are our dependencies using? If it's close enough to `4.1.12.Final`, maybe we could explicitly depend on that version, and evict other versions.

**Comment by jsancio** [comment]
@takirala regarding the error above, update finagle to the latest version.

**Comment by takirala** [comment]
Ivy Report for code up to the latest commit : [report.zip](https://github.com/dcos/cosmos/files/1163811/report.zip)

**Comment by takirala** [comment]
@cruhland Dependencies are using 4.1.10.Final (You can view the ivy report to find more details) So I am going to try and force dependency to 4.1.12.Final and see if anything breaks.

File: project/Deps.scala [hunk 3]
```diff
@@ -155,6 +158,7 @@ object V {
   val logback = "1.1.3"
   val mockito = "1.10.19"
   val mustache = "0.9.1"
+  val netty = "4.1.13.Final"
```
Comment (jsancio) [inline project/Deps.scala]: Wohoo! We are using Netty 4.1 by default!

**Comment by takirala** [comment]
[Ivy Dependency Report](https://github.com/dcos/cosmos/files/1164049/report.zip)

File: project/Deps.scala [hunk 4]
```diff
@@ -147,24 +147,23 @@ object V {
...
   val jsonSchema = "2.2.6"
-  val logback = "1.1.3"
-  val mockito = "1.10.19"
-  val mustache = "0.9.1"
+  val logback = "1.2.3"
+  val mockito = "2.8.47"
+  val mustache = "0.9.5"
   val netty = "4.1.13.Final"
   val scalaCheck = "1.13.4"
   val scalaCheckShapeless = "1.1.3"
```
Comment (takirala) [inline project/Deps.scala]: @cruhland @jsancio I am not sure what is going on [here](https://github.com/alexarchambault/scalacheck-shapeless/releases) There is tagged release for 1.1.6 but 1.1.3 is marked as the latest release. Shall I update to 1.1.6 or are we happy with 1.1.3?

File: project/Deps.scala [hunk 5]
```diff
@@ -144,23 +147,23 @@ object V {
...
   val projectVersion = "0.5.0-SNAPSHOT"
 
-  val aws = "1.11.63"
-  val bijection = "0.9.4"
-  val circe = "0.6.1"
-  val curator = "2.11.1"
-  val fastparse = "0.4.1"
-  val finch = "0.11.1"
-  val guava = "16.0.1"
+  val aws = "1.11.163"
```
Comment (takirala) [inline project/Deps.scala]: 1.11.164 is released in july and there were 0 usages in maven central. So decided to use 1.11.163. Both were released in July 17.

File: project/Deps.scala [hunk 6]
```diff
@@ -144,23 +147,23 @@ object V {
...
+  val bijection = "0.9.5"
+  val circe = "0.8.0"
+  val curator = "3.3.0"
+  val fastparse = "0.4.3"
+  val finch = "0.15.1"
```
Comment (takirala) [inline project/Deps.scala]: For the finch library, 0.16-M1 is released on June 15 but is not stable yet. So used the 0.15.1

File: project/Deps.scala [hunk 7]
```diff
@@ -144,23 +147,23 @@ object V {
...
   val scalaCheck = "1.13.4"
   val scalaCheckShapeless = "1.1.3"
   val scalaTest = "3.0.1"
-  val scalaUri = "0.4.11"
-  val slf4j = "1.7.10"
-  val twitterServer = "1.25.0"
-  val twitterUtilCore = "6.39.0"
-  val zookeeper = "3.4.6"
+  val scalaUri = "0.4.16"
+  val slf4j = "1.7.25"
```
Comment (takirala) [inline project/Deps.scala]: There were few alpha releases after this stable release. I Ignored them

File: project/Deps.scala [hunk 8]
```diff
@@ -144,23 +147,23 @@ object V {
...
+  val mustache = "0.9.5"
+  val netty = "4.1.13.Final"
   val scalaCheck = "1.13.4"
   val scalaCheckShapeless = "1.1.3"
   val scalaTest = "3.0.1"
-  val scalaUri = "0.4.11"
-  val slf4j = "1.7.10"
-  val twitterServer = "1.25.0"
-  val twitterUtilCore = "6.39.0"
-  val zookeeper = "3.4.6"
```
Comment (takirala) [inline project/Deps.scala]: This was not used anywhere. I removed it and nothing broke!

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/finch/VersionedResponsesSpec.scala [hunk 9]
```diff
@@ -82,9 +78,8 @@ object VersionedResponsesSpec {
     )
   }
 
-  def extractBody[A](result: Option[(Input, Rerunnable[Output[A]])]): A = {
-    val Some((_, eval)) = result
-    Await.result(eval.run).value
+  def extractBody[A](result: Endpoint.Result[A]): A = {
+    result.awaitValue().get.get()
```
Comment (jsancio) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/finch/VersionedResponsesSpec.scala]: Looks good but can you explain why the change from `Await.result` to `awaitValue`? Do they mention this in their release notes?

File: project/Deps.scala (hunk 8 above)
Comment (jsancio) [inline project/Deps.scala]: Sounds good. Curator depends on zookeeper. Can you double check that the one that we bring in is the same or later than the one Cosmos uses now? Ideally we should bring the latest.

I don't remember the details but I seem to remember @cruhland having to deal with our version of curator/zookeeper.

Comment (takirala) [inline project/Deps.scala]: @jsancio @cruhland 

According to [Curator Home page](http://curator.apache.org/) : 
The are currently two released versions of Curator, 2.x.x and 3.x.x:
Curator 2.x.x - compatible with both ZooKeeper 3.4.x and ZooKeeper 3.5.x
Curator 3.x.x - compatible only with ZooKeeper 3.5.x and includes support for new features such as dynamic reconfiguration, etc.

And according to [Zookeeper release page](http://zookeeper.apache.org/releases.html) all the 3.x.x releases either alpha or beta. Hence I am using the latest version of curator 2.x.x which is `2.12.0` (released March 17) and the cosmos dependencies use zookeper `3.4.8` (As `3.4.9` is the latest, so decided not to evict the `3.4.8` version)

Comment (jsancio) [inline project/Deps.scala]: Correct! DC/OS uses ZooKeeper 3.4.x so we need the latest Curator 2.x.x.

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/finch/VersionedResponsesSpec.scala (hunk 9 above)
Comment (takirala) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/finch/VersionedResponsesSpec.scala]: @jsancio  The type of `result` used to be `Option[(Input, Rerunnable[Output[A]])]` which is now changed to `result: Endpoint.Result[A]`. As I don't have a rerunnable any more, I was not able to get a `Future` to pass to `Await.result`. So I looked through the API documentation and found the `awaitValue()` which is exactly what we are doing. As we are not returning the future and basically waiting in a synchronous manner, I thought this is a right way to do it. There is a new class called `EndpointResult` in `io.finch` introduced in the latest version of `finch-core 0.15.1` that has method.

Comment (takirala) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/finch/VersionedResponsesSpec.scala]: @cruhland @jsancio Just checked the release notes of a previous version and it talks about something similar https://github.com/finagle/finch/releases/tag/0.12.0

Shall I update the method usage to `result.awaitValueUnsafe().get` as what we were/are doing is an unsafe (but acceptable) op?

Comment (jsancio) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/finch/VersionedResponsesSpec.scala]: Yes. Let's use `awaitValueUnsafe` in our tests.
